<a href="https://colab.research.google.com/github/1120rayna-cmyk/first-project/blob/main/Gaia_Star_Data_Cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
%pip -q install astroquery astropy pandas numpy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.1/11.1 MB 74.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 33.3 MB/s eta 0:00:00


In [2]:
import pandas as pd
import numpy as np

from astroquery.gaia import Gaia
from astropy.table import Table

print("All libraries imported successfully!")

In preparation for Gaia DR4, the Gaia archive is in evolution. Unfortunately, it may be unstable at times and particular types of queries may time out. Please consider registering for a user account (https://www.cosmos.esa.int/web/gaia-users/register). For questions or advice, please contact the Gaia helpdesk (https://www.cosmos.esa.int/web/gaia/gaia-helpdesk).
All libraries imported successfully!


In [4]:
from astroquery.gaia import Gaia

Gaia.TIMEOUT = 60

query = """
SELECT TOP 10 source_id, ra, dec
FROM gaiadr3.gaia_source
"""

job = Gaia.launch_job(query)
test_results = job.get_results()

print("Success! Rows received:", len(test_results))
print(test_results)


Success! Rows received: 10
    source_id              ra                dec        
                          deg                deg        
------------------ ------------------ ------------------
137333365999306624 45.003085437246824  35.31421814401729
137337626604785664  45.24158172830418  35.38207514649842
137339486325672704  45.10751729369523  35.33991028047406
137339928707256960 45.070040691646035  35.38624585109338
137340375383849472  45.14766646032983 35.377284157661705
137340718981261568  45.24663443870708  35.41248285443802
137340723276237440 45.252041556600155 35.403772818047784
137340890779951616  45.19481057433486  35.41212120144094
137341238672637440 45.157539392099494  35.39417273345611
137341479190062848 45.102734564771694  35.42273751135722


In [5]:
query = """
SELECT TOP 10000
    source_id,
    ra,
    dec,
    phot_g_mean_mag,
    phot_bp_mean_mag,
    phot_rp_mean_mag,
    parallax,
    pmra,
    pmdec
FROM gaiadr3.gaia_source
"""

job = Gaia.launch_job(query)
results = job.get_results()

print("Query successful!")
print("Rows:", len(results))
print("Columns:", results.colnames)

Query successful!
Rows: 10000
Columns: ['source_id', 'ra', 'dec', 'phot_g_mean_mag', 'phot_bp_mean_mag', 'phot_rp_mean_mag', 'parallax', 'pmra', 'pmdec']


In [6]:
df = results.to_pandas()

# Keep a copy of the original data
raw_df = df.copy()

print("Dataset shape:", df.shape)

display(df.head(10))

Dataset shape: (10000, 9)


,source_id,ra,dec,phot_g_mean_mag,phot_bp_mean_mag,phot_rp_mean_mag,parallax,pmra,pmdec
0,137333232852897664,44.977046,35.290590,21.035749,20.779573,20.848816,NaN,NaN,NaN
1,137337733979509504,45.253423,35.383471,19.350798,20.706076,18.320164,1.104897,1.577865,-3.169335
2,137338867850377216,45.137320,35.358457,19.363752,19.712234,19.032219,-0.366980,-0.143066,-0.030483
3,137339378951405952,45.028506,35.341545,18.498632,19.416975,17.567169,1.090537,-4.295335,5.426279
4,137339619469588480,45.120983,35.356513,20.176880,21.046318,19.091557,0.194413,-1.119759,-3.517083
5,137339619469589376,45.124621,35.357478,19.597198,19.931034,19.195433,0.192686,-0.491941,0.002397
6,137339791268284672,45.052265,35.362368,20.157501,20.966326,18.977390,1.457261,6.923891,-6.646016
7,137340482758073856,45.163256,35.396202,18.263525,NaN,NaN,0.111620,-0.725624,-1.139931
8,137340551477551360,45.222906,35.374810,14.576084,14.991784,13.985226,1.391977,2.814384,-7.654630
9,137340551477552128,45.214661,35.371202,19.565956,20.429516,18.680265,0.539779,-0.372351,-0.433062


In [7]:
df.info()

print("\nData types:")
print(df.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   source_id         10000 non-null  int64  
 1   ra                10000 non-null  float64
 2   dec               10000 non-null  float64
 3   phot_g_mean_mag   9989 non-null   float32
 4   phot_bp_mean_mag  9854 non-null   float32
 5   phot_rp_mean_mag  9881 non-null   float32
 6   parallax          8736 non-null   float64
 7   pmra              8736 non-null   float64
 8   pmdec             8736 non-null   float64
dtypes: float32(3), float64(5), int64(1)
memory usage: 586.1 KB

Data types:
source_id             int64
ra                  float64
dec                 float64
phot_g_mean_mag     float32
phot_bp_mean_mag    float32
phot_rp_mean_mag    float32
parallax            float64
pmra                float64
pmdec               float64
dtype: object


In [8]:
missing_report = pd.DataFrame({
    "Missing Count": df.isnull().sum(),
    "Missing Percentage": (df.isnull().mean() * 100).round(2)
})

display(missing_report)

,Missing Count,Missing Percentage
source_id,0,0.00
ra,0,0.00
dec,0,0.00
phot_g_mean_mag,11,0.11
phot_bp_mean_mag,146,1.46
phot_rp_mean_mag,119,1.19
parallax,1264,12.64
pmra,1264,12.64
pmdec,1264,12.64


In [9]:
duplicate_count = df["source_id"].duplicated().sum()

print("Duplicate source IDs:", duplicate_count)
print("Total rows:", len(df))
print("Unique source IDs:", df["source_id"].nunique())

Duplicate source IDs: 0
Total rows: 10000
Unique source IDs: 10000


In [10]:
# Check for infinite numerical values
numeric_df = df.select_dtypes(include=np.number)

print("Infinite values in each column:")
print(np.isinf(numeric_df).sum())

# Check coordinate ranges
invalid_ra = df["ra"].notna() & ~df["ra"].between(0, 360)
invalid_dec = df["dec"].notna() & ~df["dec"].between(-90, 90)

print("\nInvalid right ascension values:", invalid_ra.sum())
print("Invalid declination values:", invalid_dec.sum())

Infinite values in each column:
source_id           0
ra                  0
dec                 0
phot_g_mean_mag     0
phot_bp_mean_mag    0
phot_rp_mean_mag    0
parallax            0
pmra                0
pmdec               0
dtype: int64

Invalid right ascension values: 0
Invalid declination values: 0


In [11]:
df_clean = df.copy()

# Replace infinite values with missing values
df_clean.replace([np.inf, -np.inf], np.nan, inplace=True)

# Remove rows missing essential information
df_clean.dropna(
    subset=["source_id", "ra", "dec"],
    inplace=True
)

# Remove invalid coordinates
valid_ra = df_clean["ra"].between(0, 360)
valid_dec = df_clean["dec"].between(-90, 90)

df_clean = df_clean[valid_ra & valid_dec].copy()

# Remove duplicate source IDs
df_clean.drop_duplicates(
    subset=["source_id"],
    keep="first",
    inplace=True
)

print("Original dataset shape:", df.shape)
print("Cleaned dataset shape:", df_clean.shape)
print("Remaining duplicate IDs:",
      df_clean["source_id"].duplicated().sum())

Original dataset shape: (10000, 9)
Cleaned dataset shape: (10000, 9)
Remaining duplicate IDs: 0


In [12]:
print("Missing values after cleaning:")
display(df_clean.isnull().sum().to_frame("Missing Count"))

print("\nFinal data types:")
print(df_clean.dtypes)

print("\nFirst 10 cleaned records:")
display(df_clean.head(10))

Missing values after cleaning:


,Missing Count
source_id,0
ra,0
dec,0
phot_g_mean_mag,11
phot_bp_mean_mag,146
phot_rp_mean_mag,119
parallax,1264
pmra,1264
pmdec,1264



Final data types:
source_id             int64
ra                  float64
dec                 float64
phot_g_mean_mag     float32
phot_bp_mean_mag    float32
phot_rp_mean_mag    float32
parallax            float64
pmra                float64
pmdec               float64
dtype: object

First 10 cleaned records:


,source_id,ra,dec,phot_g_mean_mag,phot_bp_mean_mag,phot_rp_mean_mag,parallax,pmra,pmdec
0,137333232852897664,44.977046,35.290590,21.035749,20.779573,20.848816,NaN,NaN,NaN
1,137337733979509504,45.253423,35.383471,19.350798,20.706076,18.320164,1.104897,1.577865,-3.169335
2,137338867850377216,45.137320,35.358457,19.363752,19.712234,19.032219,-0.366980,-0.143066,-0.030483
3,137339378951405952,45.028506,35.341545,18.498632,19.416975,17.567169,1.090537,-4.295335,5.426279
4,137339619469588480,45.120983,35.356513,20.176880,21.046318,19.091557,0.194413,-1.119759,-3.517083
5,137339619469589376,45.124621,35.357478,19.597198,19.931034,19.195433,0.192686,-0.491941,0.002397
6,137339791268284672,45.052265,35.362368,20.157501,20.966326,18.977390,1.457261,6.923891,-6.646016
7,137340482758073856,45.163256,35.396202,18.263525,NaN,NaN,0.111620,-0.725624,-1.139931
8,137340551477551360,45.222906,35.374810,14.576084,14.991784,13.985226,1.391977,2.814384,-7.654630
9,137340551477552128,45.214661,35.371202,19.565956,20.429516,18.680265,0.539779,-0.372351,-0.433062


Error: Runtime no longer has a reference to this dataframe, please re-run this cell and try again.


# Task 3: Gaia Dataset — Data Quality Issues and Cleaning Approaches

## 1. Dataset Overview

I queried the Gaia Data Release 3 (DR3) catalogue using Python and the `astroquery` library. The dataset contains 10,000 star records and 9 columns: `source_id`, `ra`, `dec`, `phot_g_mean_mag`, `phot_bp_mean_mag`, `phot_rp_mean_mag`, `parallax`, `pmra`, and `pmdec`.

I converted the query results into a Pandas DataFrame for data inspection and cleaning.

## 2. Missing Values

I inspected missing values using Pandas `isnull().sum()` and calculated their percentages.

The following missing values were identified:

* `phot_g_mean_mag`: 11 missing values (0.11%).
* `phot_bp_mean_mag`: 146 missing values (1.46%).
* `phot_rp_mean_mag`: 119 missing values (1.19%).
* `parallax`: 1,264 missing values (12.64%).
* `pmra`: 1,264 missing values (12.64%).
* `pmdec`: 1,264 missing values (12.64%).

The `source_id`, `ra`, and `dec` columns contained no missing values.

**Approach:** I removed rows with missing source IDs or celestial coordinates, as these fields are essential for identifying and locating stars. Missing optional measurements were retained as NaN because replacing them with arbitrary values could introduce inaccuracies into the astronomical data.

## 3. Duplicate Source IDs

I checked duplicate records using Pandas `duplicated()`.

**Result:** Zero duplicate source IDs were found in the original dataset.

**Approach:** I used `drop_duplicates(subset=["source_id"], keep="first")` to ensure that each source ID appeared only once. Since no duplicates were present, no rows were removed by this operation.

## 4. Invalid and Infinite Values

I checked the numerical columns for positive and negative infinity and validated the celestial coordinate ranges.

The results were:

* Infinite values: zero in every numerical column.
* Invalid right ascension values: 0.
* Invalid declination values: 0.

Right ascension was checked against the range 0–360 degrees, while declination was checked against the range -90 to +90 degrees.

**Approach:** Infinite values were replaced with NaN, and records with invalid coordinates would be removed. However, no infinite values or invalid coordinates were found in this dataset.

## 5. Data Types

I inspected the column data types using Pandas `info()` and `dtypes`.

* `source_id`: int64
* `ra`, `dec`, `parallax`, `pmra`, and `pmdec`: float64
* `phot_g_mean_mag`, `phot_bp_mean_mag`, and `phot_rp_mean_mag`: float32

These data types are appropriate for storing source identifiers and numerical astronomical measurements. No unnecessary type conversions were required.

## 6. Final Cleaning Results

The original dataset contained 10,000 rows and 9 columns. After cleaning, the dataset still contained 10,000 rows and 9 columns.

No rows were removed because the dataset contained no duplicate source IDs, invalid coordinates, or missing essential fields.

Missing optional measurements remained as NaN. This was intentional, as the objective was to preserve the available astronomical information without introducing artificial measurements.

## 7. Conclusion

The Gaia DR3 dataset was queried, inspected, and cleaned using Python, Astroquery, Astropy, Pandas, and NumPy. Missing values, data types, duplicate identifiers, infinite values, and coordinate ranges were examined. Missing essential values and invalid coordinates were handled through appropriate filtering, infinite values were converted to NaN, and duplicate source IDs were removed where applicable.

The final dataset contains 10,000 unique star records and 9 columns. The cleaning process preserved valid observations while avoiding unnecessary data loss or unsupported imputation of missing astronomical measurements.


In [13]:
df_clean.to_csv("gaia_cleaned_stars.csv", index=False)

print("Cleaned dataset saved!")

Cleaned dataset saved!


In [14]:
from google.colab import files
files.download("gaia_cleaned_stars.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>